### DISCUSSION TITLE
Can I use poethepoet to execute commands during "python3 -m build"

### DISCUSSION THREAD

**Discussion by ssuchter** [nat-n/poethepoet#140] [Q&A]
I'm working on a project that has this in pyproject.toml:

```
...
[tool.poetry.dependencies]
python = ">=3.7"
ply = "^3.11"
setuptools = ">=65.6.3"

[build-system]
requires = ["poetry-core>=1.0.0"]
build-backend = "poetry.core.masonry.api"
```

I'm wanting to run some arbitrary code during that build process (I want to autogenerate some .py files that go into the package) 

Can I use poethepoet to do this? I tried using this:
```
[tool.poe.poetry_hooks]
pre_build  = "mkdir helloworld"
```

but it didn't seem to do anything. Is this a supported use case?

**Comment by nat-n** [comment]
Hi @ssuchter,

It looks like you need to use the [Poe the Poet poetry plugin](https://poethepoet.natn.io/poetry_plugin.html), but you're missing a couple of steps.

Firstly you need to add the Poe the Poet plugin to poetry:

```sh
poetry self add 'poethepoet[poetry_plugin]'
```

Then you need to define the task you want to have run as part of the build process:

```toml
[tool.poe.tasks]
_pre_build = "mkdir helloworld"
```

Then you can reference that task from the pre_build hook:

```toml
[tool.poe.poetry_hooks]
pre_build  = "_pre_build"
```

Note: the underscore prefix on the task name is optional, but suggested to make that task "private" in the sense that it doesn't appear in the generated documentation (when you run `poe` without any tasks) and can't be run directly.

**Comment by ssuchter** [comment]
This is very helpful, thank you!

The 'poetry self add' step - is it possible to declare that poetry should
have poethepoet as a plugin in the pyproject.toml file, obviating this
step? Or must this be performed imperatively before the 'poetry build' or
'python -m build' steps?

**Reply by nat-n** [reply]
I'm afraid that's a limitation of how poetry's plugin system works, the plugin has to be installed into the poetry tool's own venv, rather than that of the project. [See poetry docs](https://python-poetry.org/docs/master/plugins/#using-plugins)

Also I'm not sure what the `build` module is in your context, but I'd guess that running `python3 -m build` is not the same thing as `poetry build` so won't necessarily pick up the hooks from the plugin.

**Comment by ssuchter** [comment]
Thanks very much to nat-n, his answers were all correct. However, I also found that one can do the exact thing I was trying to do with pure poetry, by putting this in pyproject.toml:

[tool.poetry.build]
script = "prebuild.py"
